File: Lexicographical-PGs/Lexicographical-PG.py
import math
import time

def input_from_user() -> str:

    while True: # Continues looping until valid word is retrieved from user

        word : str = input("\nEnter a word (all characters are acceptable): ")
        word = word.strip()

        if len(word) == 0:
            print("\nWord must not be left empty.")

        else:
            break

    return word

def initialise_letters (
        letters: dict[str, int],
        word : str
    ) -> None:

    for letter in word:

        if letter not in letters: # Add a new entry to letters if one does not exist
            letters[letter] = 1
        else:
            letters[letter] += 1 # Increment the existing entry in letters

def calculate_num_of_permutations (
        letters: dict[str, int],
        word : str
    ) -> int:

    denominator : int = 1

    for value in letters.values(): # Loop through all values in letters to account for characters that appear more that once and update the denominator accordingly
        denominator *= math.factorial(value)
    
    num_of_permutations : int = math.factorial(len(word)) // denominator # Calculate number of permutations

    return num_of_permutations

def find_new_pivot(word_list: list[str]) -> int:

    for counter in range(-1, -len(word_list) - 1, -1): # Read the list from right to left

        if counter != -len(word_list): # If it has reached the last element from right to left, stop looping to avoid out of range exception
            if word_list[counter - 1] < word_list[counter]: # Check if the next element is less than the current element, if it is then that is the new pivot 
                pivot : int = counter - 1
                return pivot
        else:
            return None # None indicates that the list is in reverse order

def find_new_successor (
        suffix : list[str],
        pivot_letter : str
    ) -> int:

    for counter in range(-1, -len(suffix) - 1, -1): # Read the list from right to left

        if suffix[counter] > pivot_letter: # Find the next element in suffix that is greater than the pivot element
            return counter

def swap_pivot_and_successor (
        pivot : int,
        successor : int,
        word_list : list[str]
    ) -> None:

    temp : str = word_list[pivot]
    
    word_list[pivot] = word_list[successor]
    
    word_list[successor] = temp

def permutation_generator (
        word : str,
        permutations : list[str],
        num_of_permutations : int
    ) -> None:

    word_list : list[str] = sorted(word)

    while len(permutations) < num_of_permutations: # Starting permutation loop

        pivot : int = find_new_pivot(word_list) # find_new_pivot can return None if word_list is in descending order

        if pivot != None:
            
            successor : int = find_new_successor(word_list[pivot + 1:], word_list[pivot])
            
            swap_pivot_and_successor(pivot, successor, word_list)
            
            word_list[pivot + 1:] = word_list[pivot + 1:][::-1] # Reverse the suffix (suffix is word_list[pivot + 1:])
        else: # If word_list is in descending order, reverse it
            word_list = word_list[::-1]
            
        permutations.append("".join(word_list))

def main() -> None:

    print("\n\033[4mLexicographical Permutation Generator (Python)\033[0m")

    word : str = input_from_user()

    permutations : list[str] = [] # Permutations are served in a set which automatically handle repeated values
        
    letters : dict[str, int] = dict() # All letters within the user's word are handled through a dictionary (map)
    
    initialise_letters(letters, word)
    
    num_of_permutations : int = calculate_num_of_permutations(letters, word)
    
    print(f"\nThis word has {num_of_permutations} permutation(s).")

    start_time : float = time.perf_counter()

    permutation_generator(word, permutations, num_of_permutations)

    end_time : float = time.perf_counter()

    print(f"\nAll permutations:\n {permutations}")

    print(f"\nThe Rejection Sampling PG took {(end_time - start_time):.5f} seconds to find all permutations.\n")

if __name__ == "__main__":
    main()

File: Lexicographical-PGs/Lexicographical_PG.java
import java.math.BigInteger;
import java.util.ArrayList;
import java.util.Collections;
import java.util.HashMap;
import java.util.List;
import java.util.Scanner;

public class Lexicographical_PG {

    /**
     * Calculates the factorial of a given non-negative number
     * @param number A non-negative number of type long
     * @return The result of (number * number - 1 * number - 2)
     */
    public static BigInteger factorial(long number) {
        
        assert (number > -1) : "Number must be non-negative.";

        BigInteger total = BigInteger.ONE;

        while (number > 1) {
            total = total.multiply(BigInteger.valueOf(number));
            number -= 1;
        }

        return total;
    }

    /**
     * Uses a scanner to retrieve a valid string input from the user, checking if the input is empty, and removing leading and trailing whitespace
     * @return The validated string input from the user
     */
    public static String inputFromUser() {

        String word;
        Scanner input = new Scanner(System.in);

        while (true) { // Continues looping until valid word is retrieved from user

            System.out.printf("%nEnter a word (all characters are acceptable): ");
            word = input.nextLine().strip();

            if (word.isEmpty()) {
                System.out.printf("%nWord must not be left empty.%n");
            }

            else {
                break;
            }
        }

        input.close();

        return word;
    }

    /**
     * Initialises the letters dictionary with a character as the key and an integer as the value
     * @param word A valid word string from the inputFromUser() function
     * @return The letters dictionary containing the count of each letter from the word string
     */
    public static HashMap<Character, Integer> initialiseLetters (String word) {

        HashMap<Character, Integer> letters = new HashMap<>();
        
        for (char letter : word.toCharArray()) {

            if (!letters.containsKey(letter)) { // Add a new entry to letters if one does not exist
                letters.put(letter, 1);
            }
            else {
                letters.put(letter, letters.get(letter) + 1); // Increment the existing entry in letters
            }
        }

        return letters;
    }

    /**
     * Calculates the number of permutations that can be generated from a given word
     * @param letters A dictionary that contains the count of each letter in the 'word' variable
     * @param word A valid word from the user
     * @return A number of type BigInteger that represents the number of permutations that can be generated from a given word
     */
    public static BigInteger calculateNumOfPermutations (
        HashMap<Character, Integer> letters,
        String word
    ) {
        
        BigInteger denominator = BigInteger.ONE;

        for (int value: letters.values()) { // Loop through all values in letters to account for characters that appear more than once and update the denominator accordingly
            denominator = denominator.multiply(factorial(value));
        }

        BigInteger numOfPermutations = factorial(word.length()).divide(denominator); // Calculate number of permutations

        return numOfPermutations;

    }

    /**
     * Reads the wordList from right to left checking if the next element is lexicographically smaller than the current element,
     * if that is the case, then the next element is the pivot
     * @param wordList A list of containing the characters of the user's word
     * @return The index of a new pivot location in wordList
     */
    public static Integer findNewPivot(List<Character> wordList) {

        for (int counter = wordList.size() - 1; counter > -1; counter--) {

            if (counter != 0) {
                if (wordList.get(counter - 1) < wordList.get(counter)) {
                    return counter - 1;
                }
            }
            else {
                return null;
            }

        }
        
        return null;
    }

    /**
     * Finds a character that is lexicographically greater than the pivot in a sublist of wordList
     * @param suffix A sublist of wordList containing all characters before the pivot (reading from right to left)
     * @param pivotLetter The element at the location of the pivot point
     * @return The location of the successor as an offset from the pivot index
     */
    public static Integer findNewSuccessorOffset (
        List<Character> suffix,
        char pivotLetter
    ) {

        for (int counter = suffix.size() - 1; counter > -1; counter--) {

            if (suffix.get(counter) > pivotLetter) {
                return counter + 1;
            }
        }

        return null;
    }

    /**
     * Swaps the element at the pivot with the element at the successor in wordList
     * @param pivot Index location of the pivot
     * @param successor Index location of the successor
     * @param wordList A list of containing the characters of the user's word
     */
    public static void swapPivotAndSuccessor (
        int pivot,
        int successor,
        List<Character> wordList
    ) {

        char temp = wordList.get(pivot);

        wordList.set(pivot, wordList.get(successor));

        wordList.set(successor, temp);
    }

    /**
     * A function that generates permutations lexicographically
     * @param word A valid word from the user
     * @param permutations A list which is appended to with new permutations
     * @param numOfPermutations The number of permutations that can be created with the user's word
     */
    public static void permutationGenerator (
        String word,
        List<String> permutations,
        BigInteger numOfPermutations
    ) {

        List<Character> wordList = new ArrayList<>();

        for (char letter : word.toCharArray()) {
            wordList.add(letter);
        }

        wordList.sort(null);

        while (BigInteger.valueOf(permutations.size()).compareTo(numOfPermutations) < 0) {

            Integer pivot = findNewPivot(wordList);

            if (pivot != null) {
                
                Integer successorOffset = findNewSuccessorOffset(wordList.subList(pivot + 1, wordList.size()), wordList.get(pivot));

                int successor = pivot + successorOffset;

                swapPivotAndSuccessor(pivot, successor, wordList);

                Collections.reverse(wordList.subList(pivot + 1, wordList.size()));

            }
            else {
                Collections.reverse(wordList);
            }

            StringBuilder newWord = new StringBuilder();

            for (char letter : wordList) {
                newWord.append(letter);
            }

            permutations.add(newWord.toString());
            
        }
    }

    public static void main(String[] args) {
        
        System.out.printf("%n\u001B[4mLexicographical Permutation Generator (Java)\u001B[0m%n");

        System.out.printf("%nWarning: Words with many unique characters grow permutation counts factorially. Very long or highly varied input may take a long time and user significant memory.%n");

        String word = inputFromUser();

        List<String> permutations = new ArrayList<>(); // Permutations are served as a list

        HashMap<Character, Integer> letters = initialiseLetters(word); // All letters within the user's word are handled through a dictionary (map)

        BigInteger numOfPermutations = calculateNumOfPermutations(letters, word);

        System.out.printf("%nThis word has %d permutation(s).%n", numOfPermutations);

        long startTime = System.nanoTime();

        permutationGenerator(word, permutations, numOfPermutations);

        long endTime = System.nanoTime();

        System.out.printf("%nAll permutations:%n%s%n", permutations);

        double durationInSecond = (endTime - startTime) / 1000000000.0;

        System.out.printf("%nThe Lexicographical PG took %.5f seconds to find all permutations.%n%n", durationInSecond);
    }
}
